### DISCUSSION TITLE
V9: Custom Component Recipe

### DISCUSSION THREAD

**Discussion by blendsdk** [microsoft/fluentui#26689] [Q&A] [answered]
Hi,

Please provide some examples of how we should create custom components the V9 way? (ComponentProps, Slots,renderers,.....)
I have created several custom components (some complex and large) I would like to migrate these to v9.

Thanks.

**Comment by micahgodbolt** [comment, marked as answer]
For downstream components built on v9, there is no hard requirement that you use component props, slots etc. These are conventions that we are using to keep our components as customizable as possible. 

As for how we build v9 controls, one way to see our current best practice is to clone our repo and use the `yarn create-package` and  `yarn create-component` commands. As we refine our approach, we move our best practices into those templates.

**Comment by blendsdk** [comment]
@micahgodbolt, thank you for your answer. I have been studying the react-components code in the hope of creating and migrating my v8 components to v9, following the MS style and standards.  I would like to know why several of the functions (also in the template) are postfixed by `_unstable`?
What is the significance of this?

**Reply by micahgodbolt** [reply]
Yes, we are currently using a model where state, render, and styles are all their own hooks that get combined together at component export. We are looking at ways that users can recompose those hooks to create component variants of their own. But we aren't 100% settled on the API yet, so while the hooks are there, we don't want anyone taking a hard dependency on them yet.

**Reply by blendsdk** [reply]
I must say I quite like this approach. It keeps the inner workings of a component logically separated. One nice-to-have would be to be able to restrict the type of the components that is accepted by `children` to be of a certain type. I am redesigning my v8 `AppBar`, `AppBarLauncher` and the `AppBarPanel` components with this approach. It opens lot of possibilities for composition and a plug-n-play behavior. 

Would it be possible to get a heads-up when this api in changed?

**Reply by micahgodbolt** [reply]
We will certainly make a large announcement when we do have a final recomposition API. Our current focus right now is just getting more controls out the door.

**Reply by blendsdk** [reply]
@micahgodbolt. Thank you. Any idea when you except the recomposition API to be stable? Do you except major changes?

**Reply by micahgodbolt** [reply]
no hard plans yet. maybe this year? Depends on how components go and the demand for better recomposition. We don't expect it to be a major release due to existing API being unstable, but the final API will probably look very different than our existing (or be an abstraction layer on top of the existing)

**Reply by blendsdk** [reply]
Thank you. I guess for now I have to continue using this API. I hope you do not change it too much. Once you get the hang of it, it is very intuitive to work with.
